### ISSUE TITLE
On CPU, vectorized float tensor addition might be slower than unvectorized float tensor addition

### ISSUE THREAD

**Issue by imaginary-person** [pytorch/pytorch#60202]
## 🐛 Bug

`ATEN_CPU_CAPABILITY=default` is currently faster than `ATEN_CPU_CAPABILITY=avx2` for the addition of two `float` tensors, when multiple threads are used (for 16 threads or more on a machine with 32 physical cores & 384 GB RAM, AVX2 is slower, or gives no performance boost over unvectorized code) on many-core machines. With a single thread, AVX2 is only around 10% faster.

Disabling HyperThreading helps both of them, so I'm unable to figure out what factor is responsible for the single-threaded AVX2 performance only being slightly faster than the unvectorized single-threaded performance.

## To Reproduce

1. Run this Python script with different `ATEN_CPU_CAPABILITY` levels, while varying the number of threads:
```
import torch
import time

# 64 MB tensors
x = torch.rand(1, 256, 256, 256)
y = torch.rand(1, 256, 256, 256)

NITER = 10000

s = time.time()
for i in range(NITER):
    x + y
elapsed_float = time.time() - s
ms_per_iter_float = elapsed_float / NITER * 1000

print(ms_per_iter_float)
```
2. The results for `ATEN_CPU_CAPABILITY=default` are better than those for `ATEN_CPU_CAPABILITY=avx2`

## Expected behavior

Addition of two float tensors with `ATEN_CPU_CAPABILITY=avx2` should've been faster than `ATEN_CPU_CAPABILITY=default`.

## Environment

Tested on a dual-socket Intel Xeon Gold 6142 (32 physical cores, 384 GB RAM) & a dual-socket Intel Xeon Silver 4114 (20 physical cores, 192 GB RAM) with Ubuntu 20.04 & Ubuntu 18.04, respectively.

## Additional context

This result probably extends to other dtypes as well. And perhaps some other kernels.

cc @VitalyFedyunin @ngimel @colesbury

**Comment by imaginary-person** [comment]
According to Intel vTune, most of the time is spent in handling page faults. It also asks to check for unaligned memory accesses.

**Comment by zou3519** [comment]
This doesn't sound expected, but leaving it up to the folks cc'ed to make a judgement call

**Comment by imaginary-person** [comment]
Thanks for the update!

Intel VTune asks to use Intel Advisor for analyzing memory-access patterns, but Intel Advisor can't attach itself to a running process (so that the captured data pertains only to tensor addition) with the `--target-pid` option [when `--collect=map`](https://software.intel.com/content/www/us/en/develop/documentation/advisor-user-guide/top/command-line-interface/advisor-command-line-interface-reference/advisor-command-option-reference/target-pid.html) is used with it.

So, I'll check VTune's memory analysis instead.

**Comment by imaginary-person** [comment]
Hello @colesbury, do you have any idea as to why AVX2 vectorized addition of two float tensors is only around 10% faster than its unvectorized counterpart? Thanks!

**Comment by imaginary-person** [comment]
I do realize that there's a lot more going around with tensors in PyTorch, but if I were to create & run a [simple C++ program](https://gist.github.com/imaginary-person/80b749dcc8560ffb7294aa6b84bc7a21) with two 64 MB arrays of floats, fill them up with random values, and then benchmark (with RDTSCP, for accuracy) the time taken for adding them & storing the output in a third array of floats 1000 times, then I get the following timings with a single thread -

Vectorization-type    |     ms/iter (float)          
-----------------------|-------------------------
Unvectorized             | 4.684 ms/iteration
AVX2                         | 0.587 ms/iteration

I believe this means that there's room for optimizing tensor addition with `ATEN_CPU_CAPABILITY=avx2`, compared to `ATEN_CPU_CAPABILITY=default`.

**Comment by ajtulloch** [comment]
cc @swolchok might be interesting.

**Comment by colesbury** [comment]
It sounds like the cost is allocation/kernel zero-filling. Your C++ program is adding into the same array every iteration. If you want that in PyTorch use, `torch.add(x, y, out=out)`.

Alternatively, you can try using jemalloc as your allocator. It tends to hold onto larger allocations internally, so you don't pay the kernel zero-fill cost for every add call. (You can use `LD_PRELOAD` to run with jemalloc; it doesn't require any changes to your program).

**Comment by imaginary-person** [comment]
Thanks a lot for your help, @colesbury! That indeed seems to be the case!

I should've noticed that myself, instead of immediately resorting to VTune. 😞
But, lesson learnt! 😄 

I [revised the C++ program](https://gist.github.com/imaginary-person/80b749dcc8560ffb7294aa6b84bc7a21) to create a new array for the sum in each iteration, and the results are indeed quite similar to the Python script that uses PyTorch:

Vectorization-type    |     ms/iter (float)          
-----------------------|-------------------------
Unvectorized             | 39.276892 ms/iteration
AVX2                         | 38.755456 ms/iteration
AVX512                    |  39.073316 ms/iteration

**Comment by imaginary-person** [comment]
Even if I run the following Python script with `torch.add(x, y, out=z)`, results for `ATEN_CPU_CAPABILITY=default` & `ATEN_CPU_CAPABILITY=avx2` are quite similar to each other, with `ATEN_CPU_CAPABILITY=default` sometimes being faster than `ATEN_CPU_CAPABILITY=default` for multiple threads.
For a single thread, though, `ATEN_CPU_CAPABILITY=default` is around 12% faster, although it was around 5% faster with the first Python script.

`jemalloc` doesn't seem to make any difference with either Python script. 

I'm not sure how we can boost vectorized code's performance. 😞 

```
import torch
import time

x = torch.rand(1, 256, 256, 256)
y = torch.rand(1, 256, 256, 256)
z = torch.rand(1, 256, 256, 256)

NITER = 10000

s = time.time()
for i in range(NITER):
    torch.add(x, y, out=z)
elapsed_float = time.time() - s
ms_per_iter_float = elapsed_float / NITER * 1000

print(ms_per_iter_float)
```

**Comment by colesbury** [comment]
1. Please include the actual times, not just the relative performance
2. 64 MiB * 3 tensors = 192 MiB won't fit into L3, most accesses are from RAM
3. The limit here is memory bandwidth (to/from RAM to core), not core execution speed, so I'm not surprised that AVX (or AVX512) doesn't make a difference -- you can saturate the single core memory bandwidth with plain loads. You'll see a benefit from vectorization with smaller tensors that fit in caches or in kernels that are more compute heavy
4. It looks like in your C++ example the compiler is optimizing away the things you actually want to benchmark (once you remove the free/alloc from the inner loop).

**Comment by colesbury** [comment]
Here's my modified version of your `add_floats.cpp`: https://gist.github.com/colesbury/1c886ae02e84b87f90be1e5466efb9f1. 

I get 19.1 ms per iteration without AVX and 18.5 ms per iter with AVX. (No AVX512 support). That comes out to about `(64 MiB * 3) / (18.5 ms)  = 10.1 GiB/sec` (read+write) bandwidth which pretty closely matches the bandwidth reported by https://zsmith.co/bandwidth.php on the same machine at large (>72 MiB sizes). It's also pretty close to the PyTorch add performance.

**Comment by imaginary-person** [comment]
Thanks a lot again, @colesbury! Yes, I had chosen a tensor size that wouldn't fit into L3, so that the computation would be memory-bound. However, I didn't know how pronounced the effect would be, as all vectorization levels had quite similar figures for cache misses.

If I use 1 MB tensors in the Python script, then I do get better performance for AVX2 vs. default (single threaded - 0.121 ms/iter vs. 0.167 ms/iter, and 0.017 ms/iter vs 0.033 ms/iter for 16 threads on a 64-core machine).

Also, I just noticed that I hadn't been used RDTSCP in the code I had posted!
BTW, did you use Godbolt, etc to figure out that the compiler had optimized away what I wanted to measure, or could you tell that from experience? Thanks a lot for fixing it!

**Comment by colesbury** [comment]
> BTW, did you use Godbolt, etc to figure out that the compiler had optimized away what I wanted to measure, or could you tell that from experience? Thanks a lot for fixing it!

I copied the inner loop into a separate function to try to understand it better (with `__attribute((noinline))__` to prevent inlining) and it ran 100x slower, so my suspicion was that the original version optimized away the bulk of the loop. For these sorts of benchmarks, I might use Godbolt to prototype something, but will usually use GDB's dissasembler to verify that the assembly matches what I would expect.

**Comment by imaginary-person** [comment]
Thanks a lot for the tip, @colesbury! I'll keep that in mind while trying to reason about performance.
